### PR TITLE
Allow `SetBalance` to handle error when trying to kill acount with reference counter.

### PR DESCRIPTION
**Pull request by shawntabrizi**
The `set_balance` extrinsic could lead to an unexpected state change when trying to kill an account that cannot be killed. 

This resolves that issue, and adds a test to verify the behavior is working as expected.

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
thinking out loud, we could make a wrapper for `Dispatchable` that's used in tests, making sure that everytime `.dispatch` returns `Err`, it must be a `storage_noop!`. Not sure if there are reasonable exceptions to it though..

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -314,25 +314,32 @@ pub mod pallet {
 			let new_free = if wipeout { Zero::zero() } else { new_free };
 			let new_reserved = if wipeout { Zero::zero() } else { new_reserved };
 
-			let (free, reserved) = Self::mutate_account(&who, |account| {
-				if new_free > account.free {
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply frame/balances/src/lib.rs:318]: yes

**Comment by shawntabrizi** [comment]
If we can do it in tests, we probably should. If there are exceptions to be had, we should have some syntax to enable the exception, but most often than not, we follow this rule.

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -314,25 +314,32 @@ pub mod pallet {
...
 
-				(account.free, account.reserved)
+				(old_free, old_reserved)
 			})?;
-			Self::deposit_event(Event::BalanceSet { who, free, reserved });
+
+			// This will adjust the total issuance, which was not done by the `mutate_account`
+			// above.
+			if new_free > old_free {
+				mem::drop(PositiveImbalance::<T, I>::new(new_free - old_free));
```
Comment (ggwpez) [inline frame/balances/src/lib.rs:331]: None of the tests seem to verify the new `TotalIssuance`, since I can just out-comment both ifs.

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:331]: hm, thats not good. Ill write tests then

**Review by ggwpez — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Review by gavofyork — APPROVED** [review]

**Comment by shawntabrizi** [comment]
Cumulus CI seems to be broken. This should be fine to merge though.

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**